teps/draft-20200715-access-control-via-sa.md

```diff
@@ -0,0 +1,249 @@
+
+## Requirements
+
+Kubernetes administrators have means to enact control over which ClusterTasks, Tasks, or Pipelines are referenced
+by a PipelineRun or TaskRun.
+
+Kubernetes administrators have a means to opt into such enforcement on a granular basis (per namespace). 
+
+## Proposal
+
+An addition of a Kubernetes Validating Admission Webhook to the current Tekton Webhooks Deployment can 
+serve as the contact point for intercepting any create/update requests for TaskRun and PipelineRun and 
+then see if the ServiceAccount specified (or "default" if non specified) has permission to get/read any
+referenced ClusterTask/Task/Pipeline objects.
+
+A Kubernetes RBAC Subject Access Review (SAR) check will be executed to see if the ServiceAccount in question
+is part of a Role/RoleBinding or ClusterRole/ClusterRoleBinding that allows access to the referenced 
+ClusterTask, Task, or PipelineTask.  If access is denied, the admission webhook will prevent creation/update
+from completing per the contract specified by the Kubernetes Validating Admission Webhook API.
+
+By default, this new webhook will not perform any checks.  However, on each request it will interrogate 
+a documented environment variable (which is specified on the Tekton Webhook Deployment).  If with this environment
+variable:
+
+- "all" is specified, preceded/followed by a space, factoring in beginning and end of the value of the environment 
+variable
+- or a space delimited list of names corresponding to namespaces is provide (where "all" could be one of those 
+entries as well, hence turning the feature on for every namespace)
+
+then the new webhook will start enforcement on the incoming PipelineRun or TaskRun.
+
+A second environment variable around enforcement of cluster scoped resources, namespace scoped resources, or both,
+could also be employed if that level of configuration enablement is desired.   
+
+### User Stories (optional)
+
+As a Kubernetes cluster administrator, I want to dynamically modify Tekton's configuration such that enforcement
+of required permissions to get/read cluster scoped resources like ClusterTask from a PipelineRun or TaskRun is 
+possible, using the ServiceAccount (specified or default) as the actor in the permission check, and where 
+enablement can be done on a global basis or per namespace basis, and where enablement can vary for cluster scoped
+resources vs. namespace scoped resources.
+
+As a Kubernetes cluster administrator, I want to dynamically modify Tekton's configuration such that enforcement
+of required permissions to get/read namespace scoped resources like Task from a TaskRun or Task/Pipeline from a 
+PipelineRun is possible, where all objects are in the same namespace, using the ServiceAccount (specified or 
+default) as the actor in the permission check, and where enablement can be done on a global basis or per 
+namespace basis, and where enablement can vary for cluster scoped resources vs. namespace scoped resources. 
```

Comment:
could you update these user stories to explain more about why we have a situation where a user is able to create TaskRuns and PipelineRuns, but we want to make it so that they cannot access the referenced Task / Pipeline / ClusterTask unless the service account is modified?

I understand in principle what this proposal is suggesting but I don't understand why: if a user can create a PipelineRun / TaskRun, they can embed the Task / Pipeline spec if needed, i.e. they can ultimately run anything they want.

if this is something we want to prevent also, I think this is more motivation to tackle these more fine grained permissions as a whole, and leverage solutions like OPA that are being created for this purpose

Reply:
Sure I can add an example.  Will be part of next push.